---
author: "Centre for Population Genomics"
date: "`r Sys.time()`"
output:
  html_document:
    theme: simplex
    toc: true
    code_download: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  TITLE: "TOB-WGS concordance of SNPchip and WGS genotype calls"
  SNP_MT: ""
  WGS_MT: ""
  SAMPLES_TO_KEEP: ""
  CHROM_TO_KEEP: ""
  RES_SAMPLES: ""
  CPU: ""
title: "`r paste(params$TITLE)`"
---

```{r knitr_opts, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE, message = TRUE)
```

```{r load_pkgs}
library(dplyr, include.only = c("select", "mutate", "tibble", "bind_rows", "pull"))
library(DT, include.only = c("datatable", "formatCurrency", "formatStyle", "styleColorBar"))
library(glue, include.only = "glue")
library(htmltools, include.only = c("tags", "div"))
library(jsonlite, include.only = "fromJSON")
library(purrr, include.only = c("map", "set_names"))
library(readr, include.only = c("write_tsv", "read_tsv", "cols"))
library(readxl, include.only = "read_excel")
library(reticulate, include.only = c("use_condaenv", "py"))
library(rlang, include.only = c("is_integer", "is_double"))
library(tidyr, include.only = "unnest")
library(tidyselect, include.only = "vars_select_helpers")
use_condaenv("hailr", required = TRUE)
```

```{r read_params, eval=FALSE, echo=FALSE}
# for interactive debugging
# params <- rmarkdown::yaml_front_matter("concordance.Rmd")$params
# d <- "data"
d <- "/Users/peterd/projects/tob-wgs/nogit"
params_tmp <- list(
  local = list(
    SNP_MT = glue("{d}/snpchip/processed/snpchip_grch38.mt"),
    WGS_MT = glue("{d}/bucket/test/test-v1-raw.mt"),
    SAMPLES_TO_KEEP = "samples_to_keep.tsv",
    CHROM_TO_KEEP = "chr22",
    RES_SAMPLES = "results1.tsv",
    CPU = 8
  )
)
params <- params_tmp[["local"]]
```

```{python hail_setup}
import hail as hl
hl.init(default_reference='GRCh38', master=f'local[{r.params["CPU"]}]')
```

## Analysis

```{r read_samples}
samples_to_keep <-
  params$SAMPLES_TO_KEEP |>
  read_tsv(col_names = "sample", col_types = "c") |>
  dplyr::pull(sample) |>
  c("FAKE")
```

```{python hail_densify_split_multi}
wgs = hl.read_matrix_table(r.params['WGS_MT'])
snp = hl.read_matrix_table(r.params['SNP_MT']).key_rows_by('locus', 'alleles')
wgs_samples = wgs.s.collect() # for table in the end

##---- select chromosomes ----#
valid_chrom = ['chr' + str(i) for i in range(1, 23)]
chrom_to_keep = r.params['CHROM_TO_KEEP']
if chrom_to_keep in valid_chrom:
    print(f'Filtering by contig {chrom_to_keep}')
    wgs = wgs.filter_rows(wgs.locus.contig == chrom_to_keep)
    snp = snp.filter_rows(snp.locus.contig == chrom_to_keep)

##---- select samples ----#
wgs = wgs.filter_cols(hl.literal(r.samples_to_keep).contains(wgs['s']))
samples_not_found_in_wgs = [x for x in r.samples_to_keep if x not in wgs_samples]

#---- densify and lgt_to_gt ----#
wgs = hl.experimental.densify(wgs)
wgs = wgs.annotate_entries(GT=hl.experimental.lgt_to_gt(wgs.LGT, wgs.LA))

#---- split multiallelic and keep biallelic rows ----#
wgs = hl.split_multi_hts(wgs)
wgs = wgs.filter_rows(hl.len(wgs.alleles) == 2)

#---- unphase genotypes ----#
wgs = wgs.annotate_entries(
    GT=hl.case()
    .when(wgs.GT.is_diploid(), hl.call(wgs.GT[0], wgs.GT[1], phased=False))
    .when(wgs.GT.is_haploid(), hl.call(wgs.GT[0], phased=False))
    .default(hl.missing(hl.tcall))
)
```

## Step 4: Concordance calculation

```{python hail_concordance}
global_conc, cols_conc, rows_conc = hl.concordance(snp, wgs)

# write concordance stats per sample
cols_conc.export(r.params["RES_SAMPLES"], delimiter='\t')
```

## Results

```{r concordance_results}
get_mat_stats <- function(m) {
  mat_rename <- function(m) {
    conc_names <- c("missing_variant", "missing_gt", "homref", "het", "homalt")
    colnames(m) <- glue("{conc_names}_wgs")
    rownames(m) <- glue("{conc_names}_snp")
    m
  }
  mat_conc_stats <- function(m) {
    conc_tot1 <- sum(
      m["homref_snp", "homref_wgs"],
      m["het_snp", "het_wgs"],
      m["homalt_snp", "homalt_wgs"])
    conc_tot2 <- m["homref_snp", "missing_variant_wgs"] # these are mostly in homref WGS blocks
    conc_tot <- conc_tot1 + conc_tot2
    disc_tot1 <- sum(m[3:5, 3:5]) - conc_tot1
    disc_tot2 <- m["het_snp", "missing_variant_wgs"] + m["homalt_snp", "missing_variant_wgs"]
    disc_tot <- disc_tot1 + disc_tot2

    # missing, homrefs, hets and homalts
    snp_tot <- sum(m[c(2, 3, 4, 5), ])
    wgs_tot <- sum(m[, c(2, 3, 4, 5)])
    conc_over_snp <- conc_tot / snp_tot
    conc_over_wgs <- conc_tot / wgs_tot
    disc_over_snp <- disc_tot / snp_tot
    disc_over_wgs <- disc_tot / wgs_tot
    tibble(disc_tot = disc_tot,
           conc_tot = conc_tot,
           wgs_tot = wgs_tot,
           snp_tot = snp_tot,
           `conc / wgs` = conc_over_wgs,
           `conc / snp` = conc_over_snp,
           `disc / snp` = disc_over_snp,
           `disc / wgs` = disc_over_wgs)
  }
  m |> mat_rename() |> mat_conc_stats()
}

res <- params$RES_SAMPLES |>
  read_tsv(col_types = "ccd") |>
  mutate(conc2 = map(concordance, ~fromJSON(.) |> get_mat_stats())) |>
  unnest(conc2) |>
  select(sample = s, `conc / snp`, `conc / wgs`,
         conc_tot, disc_tot,
         snp_tot, wgs_tot,
         `disc / snp`, `disc / wgs`)

cap <- tags$caption(
  div("Concordance statistics per sample."),
  div("conc: concordant; disc: discordant;"),
  div("snp/wgs: snp and wgs datasets used in hail.concordance(snp, wgs)."),
  div("snp: SNP dataset; wgs: WGS dataset."))
```

### Concordance table

```{r concordance_table}
res |>
  mutate(across(where(is.numeric), ~round(.x, 4))) |>
  datatable(rownames = FALSE,
            caption = cap,
            class = "cell-border display compact",
            filter = list(position = "top", clear = FALSE, plain = TRUE),
            extensions = c("Scroller", "KeyTable", "Buttons"),
            options = list(scroller = TRUE, scrollX = TRUE, scrollY = 700,
                           buttons = 'csv', keys = TRUE, dom = "Blfrtip")) |>
  formatStyle(res |> select(where(is_double)) |> names(),
              background = styleColorBar(c(0, 1), "lightgreen"),
              backgroundSize = "90% 90%",
              backgroundRepeat = "no-repeat", backgroundPosition = "center") |>
  formatCurrency(res |> select(where(is_integer)) |> names(),
                 currency = "", interval = 3, mark = ",", digits = 0)
```

### Sample table

```{r sample_table}
sample_tab <- function(x, caption = "Sample table", ncols = 10) {
  ord <- sub("TOB", "", x) |> as.integer() |> order()
  nrows <- ceiling(length(x) / ncols)
  d1 <- replace(matrix(NA, nrows, ncols), seq_along(x), x[ord]) |>
    as.data.frame() |>
    set_names(paste0("Col", seq_len(ncols)))
  d1 |> knitr::kable(caption = caption)
}
sample_tab(py$wgs_samples, "Samples in WGS MT.")
sample_tab(samples_to_keep, "Samples selected for concordance check.")
cat("Samples selected for concordance but not found in WGS MT:\n", paste(py$samples_not_found_in_wgs, collapse = ", "))
```
