---
author: "Centre for Population Genomics"
date: "`r Sys.time()`"
output:
  html_document:
    theme: cosmo
    #css: ../style.css
    toc: true
    code_download: true
    code_folding: show
  rmdformats::material:
    highlight: kate
params:
  title: "Concordance of SNPchip and WGS genotype calls"
title: "`r paste(params$title)`"
---

```{r knitr_opts, include=F}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r load_pkgs}
require(tidyverse)
require(fs)
require(glue)
require(here)
require(reactable)
require(downloadthis)

require(googleCloudStorageR)
require(gargle)

source("functions.R")

# conda pkgs
CONDA_ENV <- "~/conda/envs"
CROSSMAP <- glue("{CONDA_ENV}/crossmap/bin/CrossMap.py")
BCFTOOLS <- glue("{CONDA_ENV}/bioinf/bin/bcftools")
```

## Introduction

Here we're exploring genotype concordance between SNPchip and WGS data for the
TOB-WGS project. We have been provided with one multi-sample VCF file containing
genotype calls from the SNPchip data, and several single-sample GVCF files
containing genotype calls from the WGS data.

```{r vars}
project <- "cpg-tob-wgs"
date <- "2021-04-13"
dir_bucket <- here("nogit/data/bucket")
bucket_test <- glue("{project}-test")
bucket_main <- glue("{project}-main")
bucket_upload <- glue("{project}-upload")
snpchip_raw_vcf <- here("nogit/data/snpchip/onek1k_pre_imputation_genotypes.vcf.bgz")
```

### WGS GVCF files

- Let's first explore the contents of the `r bucket_test` bucket:

```{r list_contents1, eval=FALSE}
scope <-c("https://www.googleapis.com/auth/cloud-platform")
token <- gargle::token_fetch(scopes = scope)
googleCloudStorageR::gcs_auth(token = token)

test_obj <- my_gcs_list_obj(bucket_test)
main_obj <- my_gcs_list_obj(bucket_main)
upload_obj <- my_gcs_list_obj(bucket_upload)

save_obj_rds(test_obj, "test")
save_obj_rds(main_obj, "main")
save_obj_rds(upload_obj, "upload")
```

```{r list_contents2}
test_obj <- read_obj_rds("test")
main_obj <- read_obj_rds("main")
upload_obj <- read_obj_rds("upload")
test_gvcf <-
  filter(test_obj, ftype == "GVCF") %>%
  mutate(nm = basename(name),
         nm = sub(".g.vcf.gz", "", nm))
main_gvcf <- filter(main_obj, ftype == "GVCF") %>%
  mutate(nm = basename(name),
         nm = sub(".g.vcf.gz", "", nm))
upload_gvcf <- filter(upload_obj, ftype == "GVCF") %>%
  mutate(nm = basename(name),
         nm = sub(".g.vcf.gz", "", nm))
```

- Total of **`r nrow(test_obj)`** files, consisting of **`r nrow(test_gvcf)`**
  GVCF files:

```{r gvcf_list}
gvcf_tab <- test_gvcf %>%
  mutate(size = as.character(size),
         n = dplyr::row_number()) %>%
  select(n, name, size) %>%
  reactable::reactable(
    pagination = FALSE, highlight = TRUE, searchable = TRUE,
    filterable = TRUE, bordered = TRUE,
    fullWidth = FALSE,
    columns = list(
      n = colDef(minWidth = 60),
      name = colDef(minWidth = 320),
      size = colDef(minWidth = 80)
    ))

htmlwidgets::prependContent(
  gvcf_tab,
  htmltools::h2(class = "title", glue("GVCF files in {bucket_test}")))
```

```{r gvcf_file_sizes}
theme_set(theme_bw())
# ~11.7 GB of GVCFs
test_gvcf %>%
  summarise(min = fs::as_fs_bytes(min(size)),
                   max = fs::as_fs_bytes(max(size)),
                   q1 = fs::as_fs_bytes(quantile(size, 0.25)),
                   median = fs::as_fs_bytes(median(size)),
                   q3 = fs::as_fs_bytes(quantile(size, 0.75)),
                   total = fs::as_fs_bytes(sum(size)),
                   .groups = "drop") %>%
  tidyr::pivot_longer(cols = dplyr::everything()) %>%
  knitr::kable(caption = glue("{bucket_test} GVCF file size metrics."))

ggplot(test_gvcf, aes(x = size)) +
  geom_histogram(bins = 40, fill='#aec7e8', color='#1f77b4') +
  scale_x_continuous(labels = scales::comma, breaks = scales::breaks_pretty(10)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, NA)) +
  theme(panel.grid.minor = element_blank(),
        panel.grid.major = element_blank()) +
  coord_flip() +
  ggtitle(glue::glue("GVCF file size (bytes) for {nrow(test_gvcf)} samples in {bucket_test}."))

ggplot(upload_gvcf, aes(x = size)) +
  geom_histogram(bins = 40, fill='#aec7e8', color='#1f77b4') +
  scale_x_continuous(labels = scales::comma, breaks = scales::breaks_pretty(10)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, NA)) +
  theme(panel.grid.minor = element_blank(),
        panel.grid.major = element_blank()) +
  coord_flip() +
  ggtitle(glue::glue("GVCF file size (bytes) for {nrow(upload_gvcf)} samples in {bucket_upload}."))
```

- There's also a CSV file with sample mappings:

```{r read_csv_map, message=FALSE}
sample_map <-
  here("nogit/data/bucket/test/R_210315_BINKAN1_1K1KDNA_M001.csv") %>%
  readr::read_csv()

in_test <- test_gvcf %>%
  mutate(nm = basename(name),
         nm = sub(".g.vcf.gz", "", nm)) %>%
  pull(nm)

in_sample_map <- sample_map$sample.sample_name
in_sample_map[!in_sample_map %in% in_test] # TOB1521 missing from test bucket
```

- Let's plot the QC metrics in this file:

```{r csv_map_plots, fig.width=14}
sample_map %>%
  select(sample = sample.sample_name,
         freemix = raw_data.FREEMIX,
         chimeras_pct = raw_data.PCT_CHIMERAS,
         dup_pct = raw_data.PERCENT_DUPLICATION,
         insert_size_med = raw_data.MEDIAN_INSERT_SIZE,
         cov_med = raw_data.MEDIAN_COVERAGE) %>%
  tidyr::pivot_longer(cols = -sample, names_to = "variable") %>%
  ggplot(aes(x = value, y = sample)) +
  geom_point(shape = 21, colour = 'blue', fill = "#1f77b4") +
  facet_wrap(~variable, scales = "free", nrow = 1) +
  theme(
    strip.background = element_blank(),
    panel.grid.minor = element_blank()
  )
```


### SNPchip VCF file

- Let's first look at the sample names and number of SNPs:

```{r snpchip_stats, cache=TRUE}
snpchip_samplenames <-
  system(glue("{BCFTOOLS} query -l {snpchip_raw_vcf}"), intern = TRUE) %>%
  dplyr::as_tibble() %>%
  purrr::set_names("sample") %>%
  mutate(n = dplyr::row_number(),
                n = paste0(n, ")")) %>%
  select(n, sample)

snpchip_variants <-
  system(glue("{BCFTOOLS} stats {snpchip_raw_vcf} | ",
              "grep -v '^#' | ",
              "grep 'number of records'"), intern = TRUE) %>%
  str_split("\t") %>%
  purrr::pluck(1, 4)
```

- Total of **`r nrow(snpchip_samplenames)`** samples.
- Total of **`r snpchip_variants`** variants.

```{r snpchip_samplenames2}
snpchip_samplenames %>%
  dplyr::slice(c(1:2, 154:157, (n() - 5):n())) %>%
  knitr::kable()
```

- Explore genotypes:

```{r read_vcf, eval=FALSE}
snp <- readr::read_tsv(snpchip_raw_vcf,
                     col_types = c(.default = "c", POS = "d"),
                     comment = "##")
snp %>%
  select(chr = `#CHROM`, everything(), -c(QUAL, FILTER, INFO, FORMAT)) %>%
  saveRDS(here("nogit/data/snpchip/2021-04-12_tibble.rds"))
```

```{r read_vcf2, eval=FALSE}
snp <- readRDS(here("nogit/data/snpchip/2021-04-12_tibble.rds"))
var_counts <-
  snp %>%
  select(-c(chr:ALT)) %>%
  purrr::map(~as.data.frame(table(.x))) %>%
  dplyr::bind_rows(.id = "sample") %>%
  dplyr::as_tibble()

var_counts %>%
  rename(GT = .x, Count = Freq) %>%
  ggplot(aes(y = Count, x = sample)) +
  geom_point(shape = 21, colour = 'blue', fill = "#1f77b4") +
  facet_wrap(~GT, scales = "free") +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  ggtitle("GT counts per sample",
          subtitle = "Sample labels not shown on X axis for clarity.")
```

```{r gt_counts}
knitr::include_graphics("gt_counts_per_sample.png")
```
