# Collection of useful scripts

## Variant counter

[This script](count_variants.py) counts the number of rare variants (freq<5%) present in a Hail Matrix Table object provided as input.
In addition to filtering for variants that are rare, the following additional filters are applied:

* Variants that pass GATK QC filters
* biallelic SNVs (single nucleotide changes, with only two possible alleles)
* not ref-ref (_i.e._, variable within the given dataset)

To run:

```bash
analysis-runner --dataset "tob-wgs" \
    --description "count TOB rare variants" \
    --access-level "standard" \
    --output-dir "v0" \
    count_variants.py --mt-path mt/v7.mt
```

### Variant explorer

[This script](subset_variants.py) takes as input a Hail Matrix Table object and a list of genes, and outputs reduced Hail Matrix Table objects (one per gene) with only the relevant variants for downstream analyses.
For each gene, the corresponding summary Hail Table object (generated by [this pipeline](https://github.com/populationgenomics/cellregmap-pipeline/blob/main/batch.py)) is opened, and merged  with the original Matrix Table object to create multiple subsetted Matrix Table objects (one per gene) each containing only those variants.
These are then saved in the analysis bucket for further exploration.

To run:

```bash
analysis-runner --dataset "tob-wgs" \
    --description "get variants for significant genes" \
    --access-level "standard" \
    --output-dir "tob_wgs_rv/pseudobulk_rv_association" \
    subset_variants.py --input-mt-path mt/v7.mt --genes "LMNA RPS26" \
    --output-mt-prefix significant_gene_burden_max/
```

### Variant annotation

[This script](annotate_variants.py) takes as input a Hail Matrix Table object and a txt file containing some interval-level annotations, extract variant data from the mt and annotates it based on the table provided.

To run:

```bash
analysis-runner --dataset "tob-wgs" \
    --description "annotate variants with chromatin accessibility info" \
    --access-level "test" \
    --output-dir "tob_wgs_rv/" \
    annotate_variants.py --input-mt-path mt/v7.mt \
    --annotation-df-path "tob_wgs_rv/open_chromatin_annotation/predicted_l1_celltypes_avg_peaks_chr21.csv" \
    --output-ht-path open_chromatin_annotation/open_chromatin_annotated.ht
```
