---
author: "Centre for Population Genomics"
date: "`r as.character(paste(Sys.time(), Sys.timezone()))`"
output:
  html_document:
    theme: cosmo
    toc: true
    code_download: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  title: "QC report for"
  meta_tsv: "work/meta.tsv"
  namespace: "test"
  dataset: "TOB-WGS"
  new_batches: "PBMC,Pilot/bone marrow"
title: "`r paste(params$title, params$dataset)`"
editor_options: 
  chunk_output_type: console
---

```{r knitr_opts, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r load_pkgs}
library(tidyverse)
library(fs, include.only = "as_fs_bytes")
library(glue, include.only = "glue")
library(reactable)
library(htmlwidgets)
library(htmltools)
library(ggrepel, include.only = "geom_text_repel")
library(ggforce, include.only = "geom_sina")
library(scales)
library(knitr)
library(plotly, include.only = c("ggplotly", "toWebGL"))

# for GCS authentication
library(googleCloudStorageR, include.only = c("gcs_list_objects", "gcs_auth"))
library(gargle, include.only = "token_fetch")
```

```{r read_params, eval=FALSE, echo=FALSE}
# for interactive debugging
params <- rmarkdown::yaml_front_matter("qc.Rmd")$params
```

```{r funcs}
# The given observation x is considered an outlier
# (see https://statsandr.com/blog/outliers-detection-in-r/).
# Using the m parameter for cases where there are simply too many outliers; the
# larger the value of m, the fewer values are considered outliers.
is_outlier <- function(x, m = 3) {
  stopifnot(is.numeric(x), is.numeric(m))
  (x < quantile(x, 0.25, na.rm = T) - m * IQR(x, na.rm = T)) |
  (x > quantile(x, 0.75, na.rm = T) + m * IQR(x, na.rm = T))
}
```

```{r qc_plot_prep}
meta <- params$meta_tsv %>%
  read_tsv() %>%
  select(
    sample = external_id,
    project = project,
    batch = batch,
    primary_study = primary_study,
    resequencing_label = resequencing_label,
    internal_id = s,
    singleton = sample_qc.n_singleton,
    snp = sample_qc.n_snp,
    ti_tv = sample_qc.r_ti_tv,
    het_hom = sample_qc.r_het_hom_var,
    indel_ratio = sample_qc.r_insertion_deletion,
    hard_filters = hard_filters,
    qc_metrics_filters = qc_metrics_filters,
    high_quality = high_quality,
    `chr20 mean dp` = chr20_mean_dp,
    is_female,
    sex_karyotype,
    chrY_ploidy,
    chrX_ploidy,
    contamination = r_contamination,
    chimeras = r_chimera,
    duplicates = r_duplication,
    `insert size median` = median_insert_size,
    reported_sex = sex,
    age = age,
    flowcell_lane = flowcell_lane
  )
  
meta = meta %>% mutate(
  reported_sex = case_when(
    as.character(reported_sex) == "1" ~ "M", 
    as.character(reported_sex) == "2" ~ "F", 
    TRUE ~ "unknown"
  )
)

meta = meta %>% mutate(
  batch = ifelse(
    project == "giab" | project == "syndip",
    "validation",
    str_c("batch", batch %>% str_pad(2, pad="0"))
  ),
  batch = ifelse(
    resequencing_label != "-",
    resequencing_label,
    batch
  ),
  batch = ifelse(
    primary_study != "TOB",
    primary_study,
    batch
  )
)

meta = meta %>% mutate(
  sample = ifelse(
    resequencing_label != "-",
    str_c(sample, "-", resequencing_label),
    sample
  )
)

meta = meta %>% mutate(
  is_new_batch = batch %in% (str_split(params$new_batches, ",") %>% unlist),
  highlighted_batch = if_else(is_new_batch, batch, "Other samples")
)

p_data <- meta %>%
  pivot_longer(
    cols = c(
      "contamination", "chimeras", "duplicates", "insert size median",
      "singleton", "snp", "ti_tv", "het_hom", "indel_ratio",
      "chr20 mean dp"
    ),
    names_to = "metric"
  ) %>%
  group_by(metric) %>%
  mutate(
    Quality = ifelse(high_quality, "High quality", "Low quality"),
    `New batches` = ifelse(is_new_batch, "New batches", "Other batches"),
    value_round = round(value, 3),
    label = if_else(is_outlier(value) | !high_quality, glue("{sample} ({value_round})"), NA_character_)
  )

cutoffs <-
  tibble(
    metric = c("singleton", "snp", "het_hom", "chimeras", "chr20 mean dp", "duplicates", "contamination", "insert size median"),
    upper_lim = c(400000,       8000000,  3.3,       0.05,       NA,              0.3,          0.05,            NA),
    lower_lim = c(NA,           2400000,  NA,        NA,         18,              NA,           NA,              250)) %>%
  filter(metric %in% unique(p_data$metric))
```

## QC metrics

QC metrics were calculated by [Picard Tools](https://broadinstitute.github.io/picard/)
in a KCCG workflow and by a Hail-based joint-calling pipeline. Additional
files have been provided containing the reported age and sex of each sample.

```{r plot_size_setup1}
n_metrics <- length(unique(p_data$metric))
plot_width <- n_metrics * 1.2
plot_height <- n_metrics * 1
```

Distributions of several QC metrics. Bad-quality samples according to the
[joint-calling workflow cutoffs](https://github.com/populationgenomics/joint-calling/blob/main/joint_calling/filter_cutoffs.yaml)
are shown in red points.

```{r qc_plot_global, fig.width=6, fig.height=5}
theme_set(theme_bw())  # set white background
qc_plot_global <- p_data %>%
  ggplot(aes(x = "", y = value, label = sample)) +
  geom_violin(fill = "transparent", colour = "grey80", alpha = 0.4) +
  geom_sina(aes(
    group = metric,
    colour = Quality, 
    shape = `New batches`,
    # alpha = `New batches`,
    size = `New batches`
  ), seed = 42) +
  geom_hline(data = cutoffs, aes(yintercept = upper_lim), color = "red") +
  geom_hline(data = cutoffs, aes(yintercept = lower_lim), color = "blue") +
  facet_wrap(~metric, scales = "free", nrow = 2) +
  scale_y_continuous(breaks = breaks_pretty(8)) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) +
  labs(
    title = glue("QC metrics (for {length(unique(p_data$sample))} samples)."),
    subtitle = "Click on legend values to show/hide points."
  ) +
  xlab("") +
  scale_colour_manual(values = c("High quality" = "#00BFC4", "Low quality" = "#F8766D")) +
  # http://www.sthda.com/english/wiki/ggplot2-point-shapes
  scale_shape_manual(values = c("Other batches" = 20, "New batches" = 4)) +
  # scale_alpha_manual(values = c("Other batches" = 0.5, "New batches" = 1)) +
  scale_size_manual(values = c("Other batches" = 0.5, "New batches" = 1.5))

toWebGL(ggplotly(qc_plot_global))
```

Samples with a high contamination rate:

```{r}
meta %>% 
  filter(!high_quality) %>% 
  filter(str_detect(hard_filters, "contamination")) %>%
  arrange(batch, -contamination) %>% 
  mutate(contamination = str_c(
    format(contamination * 100, digits=2, nsmall=2), '%')
  ) %>% 
  select(batch, sample, contamination) %>% 
  reactable()
```

Other samples failed QC:

```{r}
meta %>% 
  filter(!high_quality) %>% 
  filter(!str_detect(hard_filters, "contamination")) %>%
  arrange(batch) %>% 
  mutate(`Sex inferred / reported` = str_c(sex_karyotype, "/", reported_sex)) %>% 
  mutate(chimeras = str_c(format(chimeras * 100, digits=2, nsmall=2), '%')) %>% 
  select(batch, sample, hard_filters, `Sex inferred / reported`, chimeras, age) %>% 
  reactable()
```

Same data grouped by batch:

```{r plot_size_setup2}
n_batches <- length(unique(p_data$highlighted_batch))
plot_width <- n_batches * 2.7
plot_height <- n_metrics * 3
```

```{r qc_plot_per_batch, fig.width=5, fig.height=8}
# add batch sample counts to facet labels
batch_count_labels <- meta %>%
  count(highlighted_batch) %>%
  mutate(facet_label = as.character(glue("{highlighted_batch} ({n})"))) %>%
  select(highlighted_batch, facet_label) %>%
  deframe()

qc_plot_per_batch <- p_data %>%
  ggplot(aes(x = "", y = value, label = sample)) +
  geom_violin(fill = "transparent", colour = "grey80", alpha = 0.4) +
  geom_sina(
    aes(
      colour = Quality, 
      group = metric, 
      size = Quality, 
      shape = Quality
    ), 
    seed = 42
  ) +
  geom_hline(data = cutoffs, aes(yintercept = upper_lim), color="red") +
  geom_hline(data = cutoffs, aes(yintercept = lower_lim), color="blue") +
  facet_grid(
    metric~highlighted_batch, scales = "free",
    labeller = labeller(highlighted_batch = as_labeller(batch_count_labels))
  ) +
  scale_y_continuous(breaks = breaks_pretty(8)) +
  scale_colour_manual(values = c("High quality" = "#00BFC4", "Low quality" = "#F8766D")) +
  scale_size_manual(values = c("High quality" = 0.5, "Low quality" = 0.8)) +
  scale_shape_manual(values = c("High quality" = 20, "Low quality" = 4)) +
  theme(
    axis.ticks.x = element_blank(),
    axis.text.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  ) +
  ggtitle(glue("QC metrics by batch (for {length(unique(p_data$sample))} samples).")) +
  xlab("")

toWebGL(ggplotly(qc_plot_per_batch))
```

Table listing all `r nrow(meta)` samples in `r n_batches` batches:

```{r}
bar_style <- function(width = 1, fill = "#e6e6e6", height = "75%", align = c("left", "right"), color = NULL) {
  # Render a bar chart in the background of the cell
  align <- match.arg(align)
  if (align == "left") {
    position <- paste0(width * 100, "%")
    image <- sprintf("linear-gradient(90deg, %1$s %2$s, transparent %2$s)", fill, position)
  } else {
    position <- paste0(100 - width * 100, "%")
    image <- sprintf("linear-gradient(90deg, transparent %1$s, %2$s %1$s)", position, fill)
  }
  list(
    backgroundImage = image,
    backgroundSize = paste("100%", height),
    backgroundRepeat = "no-repeat",
    backgroundPosition = "center",
    color = color
  )
}

data = meta %>%
  transmute(
    n = row_number(),
    sample,
    batch,
    `sex inferred/reported` = str_c(sex_karyotype, "/", reported_sex),
    chrY_ploidy = chrY_ploidy,
    chrX_ploidy = chrX_ploidy,
    `chr20 mean DP` = `chr20 mean dp`,
    contamination = contamination,
    `hard filters` = str_replace(hard_filters, "\\[\\]", "") %>% str_replace("\\[\"", "") %>% str_replace("\"\\]", ""),
    `qc metrics filters` = str_replace(qc_metrics_filters, "\\[\\]", "") %>% str_replace("\\[\"", "") %>% str_replace("\"\\]", ""),
    `high quality` = high_quality,
    age,
    `indel ratio` = indel_ratio,
    `het/hom` = het_hom,
    `ti/tv` = ti_tv,
    snp = snp,
    singleton = singleton,
    `insert size median` = `insert size median`,
    duplicates = duplicates,
    chimeras = chimeras
  )

data %>%
  reactable(
    pagination = FALSE, highlight = TRUE, height = 500, searchable = TRUE,
    filterable = TRUE, bordered = TRUE, fullWidth = FALSE,
    columns = list(
      n = colDef(),
      batch = colDef(),
      sample = colDef(),
      `sex inferred/reported` = colDef(
        style = function(value) {
          list(background = ifelse(value != "XX/F" & value != "XY/M", 'red', 'white'))
        }
      ),
      chrY_ploidy = colDef(
        style = function(value) {
          bar_style(width = value / max(data$chrY_ploidy), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2)
      ),
      chrX_ploidy = colDef(
        style = function(value) {
          bar_style(width = value / max(data$chrX_ploidy), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2)
      ),
      `chr20 mean DP` = colDef(
        style = function(value) {
          bar_style(width = value / max(data$`chr20 mean DP`), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 1, suffix = "x")
      ),
      contamination = colDef(
        style = function(value) {
          bar_style(width = value / max(data$contamination), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2, percent = TRUE)
      ),
      `hard filters` = colDef(),
      `qc metrics filters` = colDef(),
      `high quality` = colDef(style = function(value) { list(background = ifelse(value, 'white', 'red')) }),
      age = colDef(
        style = function(value) {
          bar_style(width = value / max(data$age), fill = "hsl(208, 70%, 90%)")
        }
      ),
      `indel ratio` = colDef(
        style = function(value) {
          bar_style(width = value / max(data$`indel ratio`), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2)
      ),
      `het/hom` = colDef(
        style = function(value) {
          bar_style(width = value / max(data$`het/hom`), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2)
      ),
      `ti/tv` = colDef(
        style = function(value) {
          bar_style(width = value / max(data$`het/hom`), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2)
      ),
      snp = colDef(
        style = function(value) {
          bar_style(width = value / max(data$snp), fill = "hsl(208, 70%, 90%)")
        }
      ),
      singleton = colDef(
        style = function(value) {
          bar_style(width = value / max(data$singleton), fill = "hsl(208, 70%, 90%)")
        }
      ),
      `insert size median` = colDef(
        style = function(value) {
          bar_style(width = value / max(data$`insert size median`), fill = "hsl(208, 70%, 90%)")
        }
      ),
      duplicates = colDef(
        style = function(value) {
          bar_style(width = value / max(data$duplicates), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2, percent = TRUE)
      ),
      chimeras = colDef(
        style = function(value) {
          bar_style(width = value / max(data$chimeras), fill = "hsl(208, 70%, 90%)")
        },
        format = colFormat(digits = 2, percent = TRUE)
      )
    )
  )
```

## Sex ploidies

Samples with mismatching sex, sex aneuploidy or not enough data to unambigously infer sex:

```{r}
sex_data = meta %>%
  mutate(
    sex_mismatched = ifelse(
      (reported_sex == "F" & sex_karyotype != "XX") | 
      (reported_sex == "M" & sex_karyotype != "XY"), 
      T, 
      F
    ),
    sex_aneuploidy = !is.na(hard_filters) & str_detect(hard_filters, "sex_aneuploidy"),
    ambiguous_sex = !is.na(hard_filters) & str_detect(hard_filters, "ambiguous_sex"),
    bad_sex = (sex_mismatched | sex_aneuploidy | ambiguous_sex)
  ) %>%
  select(
    sex_mismatched,
    sex_aneuploidy,
    ambiguous_sex,
    bad_sex,
    hard_filters,
    sample,
    batch,
    `X ploidy` = chrX_ploidy,
    `Y ploidy` = chrY_ploidy,
    karyotype = sex_karyotype,
    `reported sex` = reported_sex,
    age
  ) %>%
  mutate(
    label = ifelse(
      bad_sex,
      glue("{batch}/{sample}, {karyotype}"), 
      NA_character_
    ),
    label = ifelse(
      is.na(label) | is.nan(age),
      label,
      glue("{label}, age: {age}")
    ),
    label = ifelse(
      !is.na(label) & (sex_aneuploidy | ambiguous_sex),
      glue("{label}, hard_filters: {hard_filters}"),
      label
    )
  ) %>%
  arrange(bad_sex)

sex_data %>% 
  filter(bad_sex) %>% 
  mutate(
    `sex inferred/reported` = str_c(karyotype, "/", `reported sex`)
  ) %>% 
  select(
     batch,
    sample,
    `sex inferred/reported`,
    `X ploidy`,
    `Y ploidy`,
    hard_filters
  ) %>% 
  arrange(batch) %>% 
  reactable()
```

X ploidy vs Y ploidy. Highlighted samples with mismatched reported and inferred sex, or basically anything but XX/F or XY/M.

```{r}
sex_data %>% 
  ggplot(aes(x = `X ploidy`, y = `Y ploidy`, color = `reported sex`)) +
  geom_point(size = 0.6) +
  geom_text_repel(
    aes(label = label), 
    color = "black", 
    na.rm = TRUE, 
    max.overlaps = NA,
    size = 2.3
  )
```

Y ploidy vs age. Samples with mismatched reported and inferred sex are labelled. chrY degrades with age, so low chrY ploidies for older males are expected (though not high chrX ploidies).

```{r}
sex_data %>% 
  ggplot(aes(y = `Y ploidy`, x = age, color = `reported sex`)) +
  geom_point(size = 0.6) +
  geom_text_repel(
    aes(label = label), 
    color = "black", 
    na.rm = TRUE, 
    max.overlaps = NA,
    size = 2.3
  )
```


## Lanes and flowcells

Sample counts per flowcell-lane ID:

```{r flowcell_lane_count_plot}
flowcell_meta = meta %>% 
  filter(flowcell_lane != "-") %>% 
  mutate(
    flowcell_lane_data = str_split(flowcell_lane, "\\."),
    flowcell1 = map_chr(flowcell_lane_data, 1),
    lanes1 = map_chr(flowcell_lane_data, 2),
    flowcell2 = map_chr(flowcell_lane_data, 3, .default = NA_character_),
    lanes2 = map_chr(flowcell_lane_data, 4, .default = NA_character_)
  ) %>% 
  select(-flowcell_lane_data)

flowcell_meta %>%
  count(flowcell_lane) %>%
  ggplot(aes(x = reorder(flowcell_lane, -n), y = n)) +
  geom_bar(stat = "identity", fill = "grey70") +
  geom_text(aes(label = n), hjust=-0.25) +
  xlab("Flowcell lane") +
  coord_flip()
```

## Duplication rate

Duplication rate against the flowcell and batch, grouped by batch. Lane numbers
are dropped, and for samples with multiple flowcell IDs, only the first one is
used (e.g. H2WCNDSX2.1-2-3-4.H2WCTDSX2.1-2-3-4 becomes H2WCNDSX2).

```{r plot_size_setup3}
plot_width <- 7
plot_height <- 7
```

```{r dup_rate_flowcell, fig.width=5, fig.height=4}
p_data_flowcell = p_data
  filter(flowcell_lane != "-") %>% 
  mutate(
    flowcell_lane_data = str_split(flowcell_lane, "\\."),
    flowcell1 = map_chr(flowcell_lane_data, 1)
  )
  
p_dup_rate_flowcell <- p_data_flowcell %>%
  filter(metric == "duplicates") %>%
  ggplot(aes(x = "", y = value, label = sample, group = metric)) +
  geom_violin(fill = "transparent") +
  geom_sina(aes(colour = flowcell), seed = 42) +
  geom_text_repel(color = "black", na.rm = TRUE, size = 2.3) +
  scale_y_continuous(breaks = breaks_pretty(8)) +
  theme(
    axis.ticks.x = element_blank(), axis.text.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()) +
  facet_wrap(~batch) +
  scale_color_manual(
    values = c("#89C5DA", "#DA5724", "#74D944", "#CE50CA", "#3F4921", "#C0717C", "#CBD588", "#5F7FC7",
               "#673770", "#D3D93E", "#38333E", "#508578", "#D7C1B1", "#689030", "#AD6F3B", "#CD9BCD",
               "#D14285", "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F", "#D1A33D",
               "#8A7C64", "#599861")) +
  ggtitle(glue("Duplication rate groupped by batch, colored by flowcell")) +
  xlab("") +
  ylab('Duplication rate')

toWebGL(ggplotly(p_dup_rate_flowcell))
```

Duplication rate against the flowcell and batch, grouped by flowcell.
Flowcell IDs with <= 2 samples are removed, because the violin plots  don't work for them.

```{r plot_size_setup4}
plot_width <- 8
plot_height <- 8
```

```{r fig.width=5, fig.height=4}
lanes_with_more_than_2_samples <- flowcell_meta %>%
  group_by(flowcell1) %>%
  tally %>%
  filter(n > 2) %>% pull(flowcell1) %>% unique()

p_data_flowcell %>%
  filter(metric == 'duplicates') %>%
  filter(flowcell1 %in% lanes_with_more_than_2_samples) %>%
  ggplot(aes(x = "", y = value, label = label, group = metric)) +
  geom_violin(fill = "transparent") +
  geom_sina(aes(colour = batch), seed = 42) +
  geom_text_repel(color = "black", na.rm = TRUE, size = 2.3) +
  scale_y_continuous(breaks = breaks_pretty(8)) +
  theme(axis.ticks.x = element_blank(), axis.text.x = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank()) +
  facet_wrap(~flowcell1) +
  ggtitle(glue("Duplication rate grouped by flowcell, colored by batch")) +
  xlab("") + ylab('Duplication rate')
```
