---
author: "Centre for Population Genomics"
date: "`r as.character(paste(Sys.time(), Sys.timezone()))`"
output:
  html_document:
    theme: cosmo
    toc: true
    code_download: true
    code_folding: show
  rmdformats::material:
    highlight: kate
params:
  title: "TOB-WGS SNPchip IDs"
  input: "snpchip/v1"
title: "`r paste(params$title)`"
---

```{r knitr_opts, include=F}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Introduction

We have been provided with one multi-sample VCF file containing genotype calls
from the SNPchip data. This VCF file was provided in 2021-Mar-17 along with the
original genotype call files. The VCF file was generated using GenomeStudio and
its Plink plugin. The GRCh37 human genome assembly was used.

The TOB sample IDs were originally given to the participants at the time of
blood collection. The genotype platform generated new IDs for the cohort which
were required in the Plink files (Family and Individual IDs (FID and IID)),
and a new key ID was generated by combining `FID_IID` (e.g. `1_1`, `2_2`).

## Exploration

We can map the SNPchip sample names to TOB IDs using the Excel spreadsheet
provided in 2021-Apr-15 by GWCCG.

```{r load_pkgs}
require(DT)
require(glue)
require(janitor)
require(knitr)
require(readxl)
require(sessioninfo)
require(tidyverse)
```

- Copying inputs to container:

```{r setup_inputs}
input_gcs <- params$input
snpchip_vcf_raw_gcs <- glue("{input_gcs}/onek1k_pre_imputation_genotypes.vcf.gz")
snpchip_vcf_raw <- basename(snpchip_vcf_raw_gcs)
snpchipid2tobid_excel_gcs <- glue("{input_gcs}/OneK1K_sample_IDs_2021-Apr-15.xlsx")
snpchipid2tobid_excel <- basename(snpchipid2tobid_excel_gcs)

# copy inputs to container
system(glue("gsutil cp {snpchipid2tobid_excel_gcs} {snpchip_vcf_raw_gcs} ."))
```

- Get SNPchip VCF sample names:

```{r snpchip_vcf_stats}
chip_sample_nms <-
  system(glue("bcftools query -l {snpchip_vcf_raw}"), intern = TRUE) %>%
  as_tibble_col("sample") %>%
  mutate(chip_vcf_ind = dplyr::row_number(),
         chip_vcf_ind = paste0(chip_vcf_ind, ".")) %>%
  select(chip_vcf_ind, sample)

chip_varcount <-
  system(glue("bcftools stats {snpchip_vcf_raw} | ",
              "grep -v '^#' | ",
              "grep 'number of records'"), intern = TRUE) %>%
  str_split("\t") %>%
  pluck(1, 4)
```

- SNPchip VCF stats:
  - Total of **`r nrow(chip_sample_nms)`** samples.
  - Total of **`r chip_varcount`** variants.

- We note that most of the sample names have different FID and IID from
  column 156 onwards, and the final 5 samples have a `_2` suffix:

```{r snpchip_sample_nms_slice}
chip_sample_nms %>%
  dplyr::slice(c(1:2, 154:157, (n() - 5):n())) %>%
  knitr::kable()
```

- Let's examine the sample IDs provided in the Excel spreadsheet:

```{r snpchip_id2tob_map_explore}
chip_id2tob_raw <-
  snpchipid2tobid_excel %>%
  read_excel(skip = 1) %>%
  clean_names() %>%
  mutate(x1 = person == glue("{index_id_fid}_{internal_id_iid}"),
         tob_id2 = formatC(tob_id, width = 5, format = "d", flag = "0"),
         tob_id2 = glue("TOB-{tob_id2}"),
         x2 = tob_id2 == tob_dark_id,
         tob_id_final = glue("TOB{tob_id}"))

glimpse(chip_id2tob_raw)
# is 'person' equal to 'FID_IID'?
chip_id2tob_raw %>% count(x1)
# is 'tob_dark_id' equal to TOB-0XXXX?
chip_id2tob_raw %>% count(x2) # 31 FALSE
# which x2s are FALSE?
chip_id2tob_raw %>%
  filter(!x2) %>%
  select(person, tob_dark_id, tob_id2)
# So TOB-01856 should be TOB-01857, and the rest simply miss a zero prefix.
chip_id2tob_raw %>%
  filter(!x2) %>%
  select(person, tob_dark_id, tob_id2) %>%
  mutate(tob_id3 = sub("-0", "-", tob_id2),
         x3 = tob_id3 == tob_dark_id) %>%
  filter(!x3)

# all samples in excel file correspond to a SNPchip VCF sample
table(chip_id2tob_raw$person %in% chip_sample_nms$sample)
# 51 samples in SNPchip VCF aren't in excel file
table(chip_sample_nms$sample %in% chip_id2tob_raw$person)
```

### Table of 51 missing samples

```{r dt51}
chip_sample_nms %>%
  filter(!sample %in% chip_id2tob_raw$person) %>%
  datatable(rownames = TRUE,
            caption = "Samples in SNPchip VCF but not in Excel metadata file.",
            class = "cell-border display compact",
            filter = list(position = "top", clear = FALSE, plain = TRUE),
            extensions = c("Scroller", "KeyTable", "Buttons"),
            options = list(scroller = TRUE, scrollX = TRUE, scrollY = 700,
                           buttons = 'csv', keys = TRUE, dom = "Blfrtip"))
```

-------------------------

## Session Info

```{r session_info, echo=FALSE}
si <- session_info(include_base = TRUE)
si_pl <- unclass(si$platform) %>% as_tibble() %>% t()
si_pkg <- unclass(si$packages) %>%
  as_tibble() %>%
  select(package, version = ondiskversion, datestamp = date)

si_pkg %>%
  datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
            rownames = FALSE, extensions = c("Scroller"),
            options = list(scroller = TRUE, scrollX = TRUE, scrollY = 400,
                           dom = 'frtip'))

tibble(var = rownames(si_pl),
       value = si_pl[, , drop = TRUE]) %>%
  kable(caption = "Platform information.")
```
